src/main/java/org/wildfly/security/auth/provider/oauth2/OAuth2SecurityRealm.java

```diff
@@ -0,0 +1,313 @@
+            try {
+                if (exists()) {
+                    return new NamePrincipal(getClaims().getString(configuration.principalClaimName));
+                }
+            } catch (Exception e) {
+                throw ElytronMessages.log.oauth2RealmFailedToObtainPrincipal(e);
+            }
+            return null;
+        }
+
+        @Override
+        public boolean verifyEvidence(Evidence evidence) throws RealmUnavailableException {
+            return isValidToken(introspectToken());
+        }
+
+        @Override
+        public boolean exists() throws RealmUnavailableException {
+            return getClaims() != null;
+        }
+
+        @Override
+        public AuthorizationIdentity getAuthorizationIdentity() throws RealmUnavailableException {
+            if (exists()) {
+                return new AuthorizationIdentity() {
+                    private Attributes attributes;
+
+                    @Override
+                    public Attributes getAttributes() {
+                        if (this.attributes == null) {
+                            this.attributes = toAttributes(claims);
+                        }
+
+                        return this.attributes;
+                    }
+                };
+            }
+
+            return null;
+        }
+
+        @Override
+        public SupportLevel getCredentialAcquireSupport(Class<? extends Credential> credentialType, String algorithmName) throws RealmUnavailableException {
+            return SupportLevel.UNSUPPORTED;
+        }
+
+        @Override
+        public <C extends Credential> C getCredential(Class<C> credentialType) throws RealmUnavailableException {
+            return null;
+        }
+
+        @Override
+        public SupportLevel getEvidenceVerifySupport(Class<? extends Evidence> evidenceType, String algorithmName) throws RealmUnavailableException {
+            if (BearerTokenEvidence.class.equals(evidenceType)) {
+                return SupportLevel.SUPPORTED;
+            }
+
+            return SupportLevel.UNSUPPORTED;
+        }
+
+        private JsonObject getClaims() throws RealmUnavailableException {
+            if (this.claims == null) {
+                JsonObject claims = introspectToken();
+
+                if (isValidToken(claims)) {
+                    this.claims = claims;
+                }
+            }
+
+            return this.claims;
+        }
+
+        private boolean isValidToken(JsonObject claims) {
+            return claims != null && claims.getBoolean("active", false);
+        }
+
+        private JsonObject introspectToken() throws RealmUnavailableException {
+            if (this.evidence != null) {
+                try {
+                    return introspectAccessToken(configuration.tokenIntrospectionUrl,
+                            configuration.clientId, configuration.clientSecret, this.evidence.getToken(), configuration.sslContext, configuration.hostnameVerifier);
```

Comment:
It's best to make these fields be package-private if you're going to access them directly, so that the compiler doesn't have to generate trampoline methods for you.